// typesafe-i18n/runtime/cjs/runtime/src/index.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.i18nString = exports.i18nObject = exports.initI18nObjectLoaderAsync = exports.initI18nObjectLoader = exports.i18nObjectLoaderAsync = exports.i18nObjectLoader = exports.i18n = void 0;
var core_1 = require("./core.cjs");
var util_instance_1 = require("./util.instance.cjs");
Object.defineProperty(exports, "i18n", { enumerable: true, get: function () { return util_instance_1.i18n; } });
var util_loader_1 = require("./util.loader.cjs");
Object.defineProperty(exports, "i18nObjectLoader", { enumerable: true, get: function () { return util_loader_1.i18nObjectLoader; } });
Object.defineProperty(exports, "i18nObjectLoaderAsync", { enumerable: true, get: function () { return util_loader_1.i18nObjectLoaderAsync; } });
Object.defineProperty(exports, "initI18nObjectLoader", { enumerable: true, get: function () { return util_loader_1.initI18nObjectLoader; } });
Object.defineProperty(exports, "initI18nObjectLoaderAsync", { enumerable: true, get: function () { return util_loader_1.initI18nObjectLoaderAsync; } });
var util_object_1 = require("./util.object.cjs");
Object.defineProperty(exports, "i18nObject", { enumerable: true, get: function () { return util_object_1.i18nObject; } });
var util_string_1 = require("./util.string.cjs");
Object.defineProperty(exports, "i18nString", { enumerable: true, get: function () { return util_string_1.i18nString; } });
